File: Maria Mercedes Atim/JAVASCRIPT/Clase 10/Orden.js
class Orden {
    static contadorOrdenes = 0; // Contador estático de órdenes
    static MAX_PRODUCTOS = 5;    // Máximo de productos por orden

    constructor() {
        this._idOrden = ++Orden.contadorOrdenes; // Asigna un ID único a la orden
        this.productos = []; // Arreglo para almacenar los productos de la orden
        this.contadorProductosAgregados = 0; // Contador de productos agregados a la orden
    }

    // Método para agregar un producto a la orden
    agregarProducto(producto) {
        if (this.contadorProductosAgregados < Orden.MAX_PRODUCTOS) { // Verifica si no se ha alcanzado el máximo
            this.productos.push(producto); // Agrega el producto al arreglo
            this.contadorProductosAgregados++; // Incrementa el contador
        } else {
            console.error("No se pueden agregar más productos a la orden."); // Mensaje de error si se supera el máximo
        }
    }

    // Método para calcular el total de la orden
    calcularTotal() {
        return this.productos.reduce((total, producto) => total + producto.getPrecio(), 0); // Suma los precios de los productos
    }

    // Método para mostrar la orden
    mostrarOrden() {
        console.log(`Orden ID: ${this._idOrden}`); // Muestra el ID de la orden
        this.productos.forEach(producto => { // Itera sobre los productos
            console.log(producto.toString()); // Muestra la representación en cadena de cada producto
        });
        console.log(`Total: ${this.calcularTotal()}`); // Muestra el total de la orden
    }
}

File: Java/Clase12/Clase12/src/ar/com/codesystem/ventas/Orden.java
package ar.com.codesystem.ventas;

public class Orden {

    private int idOrden;
    private Producto productos[];
    private static int contadorOrdenes; // Contador estático de órdenes creadas
    private int contadorProductos; // Contador de productos en la orden
    private static final int MAX_PRODUCTOS = 10;

    public Orden() { // Constructor de la clase Orden
        this.idOrden = ++Orden.contadorOrdenes; // Asigna un ID único a la orden
        this.productos = new Producto[Orden.MAX_PRODUCTOS]; // Inicializa el arreglo de productos
    }

    public void agregarProducto(Producto producto){ // Método para agregar un producto a la orden
        if(this.contadorProductos < Orden.MAX_PRODUCTOS){ // Verifica si hay espacio para más productos
            this.productos[this.contadorProductos++] = producto; // Agrega el producto al arreglo y aumenta el contador
        }
        else{ // Si se supera el máximo de productos
            System.out.println("Se ha superado el maximo de productos: " + Orden.MAX_PRODUCTOS + " productos."); // Mensaje de error
        }
    }

    public double calcularTotal(){ // Método para calcular el total de la orden
        double total = 0; // Variable para acumular el total
        for (int i = 0; i < this.contadorProductos; i++) { // Recorre los productos en la orden
            total += this.productos[i].getPrecio(); // Suma el precio de cada producto al total
        }
        return total; // Retorna el total calculado
    }

    public void mostrarOrden() { // Método para mostrar los detalles de la orden
        StringBuilder sb = new StringBuilder(); // Crea un StringBuilder para construir la salida por consola
        sb.append("idOrden: ").append(idOrden).append("\n"); // Agregamos el ID de la orden
        double totalOrden = calcularTotal(); // guardamos en la variable totalOrden el resultado del calculo total.
        sb.append("El total de la orden es: $").append(totalOrden).append("\n"); // Agrega el total de la orden
        sb.append("Productos de la Orden: \n"); // Encabezado para los productos
        for (int i = 0; i < this.contadorProductos; i++) { // Recorre los productos en la orden
            sb.append(this.productos[i]).append("\n"); // Agrega cada producto al StringBuilder
        }
        System.out.println(sb.toString()); // Muestra el contenido del StringBuilder
    }

}